packages/cojson/src/sync.ts

```diff
@@ -248,66 +228,101 @@ export class SyncManager {
-        peer.optimisticKnownStates.get(id) || emptyKnownState(id);
-
-      const sendPieces = async () => {
-        let lastYield = performance.now();
-        for (const [_i, piece] of newContentPieces.entries()) {
-          this.trySendToPeer(peer, piece).catch((e: unknown) => {
-            logger.error("Error sending content piece", { err: e });
-          });
+      requested.add(coValue.id);
+
+      for (const id of coValue.getDependedOnCoValues()) {
+        const entry = this.local.coValuesStore.get(id);
 
-          if (performance.now() - lastYield > 10) {
-            await new Promise<void>((resolve) => {
-              setTimeout(resolve, 0);
+        if (entry.state.type === "available") {
+          buildOrderedCoValueList(entry.state.coValue);
+        }
+      }
+
+      coValuesOrderedByDependency.push(coValue);
+    };
+
+    for (const entry of this.local.coValuesStore.getValues()) {
+      switch (entry.state.type) {
+        case "unavailable":
+          // If the coValue is unavailable and we never tried this peer
+          // we try to load it from the peer
+          if (!peer.toldKnownState.has(entry.id)) {
+            await entry.loadFromPeers([peer]).catch((e: unknown) => {
+              logger.error("Error sending load", { err: e });
             });
-            lastYield = performance.now();
           }
-        }
-      };
+          break;
+        case "available":
+          const coValue = entry.state.coValue;
+
+          // Build the list of coValues ordered by dependency
+          // so we can send the load message in the correct order
+          buildOrderedCoValueList(coValue);
+          break;
+      }
 
-      sendPieces().catch((e) => {
-        logger.error("Error sending new content piece, retrying", { err: e });
+      // Fill the missing known states with empty known states
+      if (!peer.optimisticKnownStates.has(entry.id)) {
         peer.optimisticKnownStates.dispatch({
-          type: "SET",
-          id,
-          value: optimisticKnownStateBefore ?? emptyKnownState(id),
+          type: "SET_AS_EMPTY",
+          id: entry.id,
         });
-        return this.sendNewContentIncludingDependencies(id, peer);
-      });
+      }
+    }
 
-      peer.optimisticKnownStates.dispatch({
-        type: "COMBINE_WITH",
-        id,
-        value: coValue.knownState(),
+    for (const coValue of coValuesOrderedByDependency) {
+      /**
+       * We send the load messages to:
+       * - Subscribe to the coValue updates
+       * - Start the sync process in case we or the other peer
+       *   lacks some transactions
+       */
+      peer.toldKnownState.add(coValue.id);
+      await this.trySendToPeer(peer, {
+        action: "load",
+        ...coValue.knownState(),
+      }).catch((e: unknown) => {
+        logger.error("Error sending load", { err: e });
       });
     }
   }
 
-  addPeer(peer: Peer) {
+  nextPeer: Map<PeerID, Peer> = new Map();
+
+  async addPeer(peer: Peer) {
```

Comment:
Major fix here.

Before we weren't waiting for the previous peer to completely close and we lacked support multiple subsequent reconnections.